### PR TITLE
Fix undefined behavior in the test_bitcoin tests

### PR DESCRIPTION
**Pull request by defuse**

### CODE REVIEW

File: src/test/scriptnum_tests.cpp [hunk 1]
```diff
@@ -180,18 +196,28 @@ BOOST_AUTO_TEST_CASE(operators)
...
-            RunOperators(values[i], values[j]);
-            RunOperators(values[i], -values[j]);
-            RunOperators(values[i] + values[j], values[j]);
-            RunOperators(values[i] + values[j], -values[j]);
-            RunOperators(values[i] - values[j], values[j]);
-            RunOperators(values[i] - values[j], -values[j]);
-            RunOperators(values[i] + values[j], values[i] + values[j]);
-            RunOperators(values[i] + values[j], values[i] - values[j]);
-            RunOperators(values[i] - values[j], values[i] + values[j]);
-            RunOperators(values[i] - values[j], values[i] - values[j]);
```
Comment (defuse) [inline src/test/scriptnum_tests.cpp:194]: Note that the old code was broken, indexing `values[j]` instead of `offsets[j]`.

File: src/consensus/params.h [hunk 2]
```diff
@@ -100,7 +100,7 @@ struct Params {
 
     uint256 hashGenesisBlock;
 
-    bool fCoinbaseMustBeShielded;
```
Comment (defuse) [inline src/consensus/params.h:103]: This change fixed:
```
$ src/test/test_bitcoin -t rpc_wallet_tests
Running 24 test cases...
consensus/params.h:91:8: runtime error: load of value 88, which is not a valid value for type 'bool'
```

**Comment by str4d** [comment]
The `scriptnum_test` changes might conflict with bitcoin/bitcoin#7095, which we want to backport as part of removing OpenSSL. Could you backport that commit and see if it conflicts, and rework your changes on top if it does?

**Comment by str4d** [comment]
(If the UB is still present after bitcoin/bitcoin#7095, then making this PR be a commit on top of that backport will make it easier to upstream.)

**Comment by daira** [comment]
The problem with applying #7095 is that the `CScriptNum` implementation from Bitcoin 0.10.0 already had undefined behaviour that was fixed in #4454. If we apply that fix also to the `CScriptNum10` implementation, then we would essentially be comparing the current `CScriptNum` with itself, which is not a useful test.

I believe a more useful test would be to compare against fixed test vectors that we can compute manually (or in Python).

**Comment by daira** [comment]
Most of this was fixed in #4740; I will rebase the remainder.

File: src/test/scriptnum_tests.cpp (hunk 1 above)
Comment (daira) [reply src/test/scriptnum_tests.cpp]: In #4740 I assumed the code in the body was as intended, and changed the loop to be over `values`. Either interpretation is correct, but I think `values` provides slightly better coverage of corner cases (I also checked that the test would pass with `offsets`).

**Review by daira — APPROVED** [review]
utACK; needs one more review.

**Review by str4d — APPROVED** [review]
utACK

**Comment by str4d** [comment]
@zkbot r+
